File: apps/web/src/store/matchmaking-store.ts
'use client';

import { create } from 'zustand';
import { apiFetch } from '../lib/api';

type MatchmakingStatus = 'idle' | 'queued' | 'matched';

interface MatchmakingStore {
  status: MatchmakingStatus;
  waitSeconds: number;
  queueCount: number;
  matchedGameId: string | null;
  error: string | null;
  _idleRetries: number;

  joinQueue: () => Promise<void>;
  leaveQueue: () => Promise<void>;
  pollStatus: () => Promise<void>;
  reset: () => void;
}

export const useMatchmakingStore = create<MatchmakingStore>((set, get) => ({
  status: 'idle',
  waitSeconds: 0,
  queueCount: 0,
  matchedGameId: null,
  error: null,
  _idleRetries: 0,

  joinQueue: async () => {
    try {
      const data = (await apiFetch('/api/v1/matchmaking/queue', { method: 'POST' })) as {
        queueCount?: number;
      };
      set({
        status: 'queued',
        waitSeconds: 0,
        queueCount: data.queueCount ?? 0,
        error: null,
        matchedGameId: null,
        _idleRetries: 0,
      });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : 'Failed to join queue' });
    }
  },

  leaveQueue: async () => {
    try {
      await apiFetch('/api/v1/matchmaking/queue', { method: 'DELETE' });
      set({ status: 'idle', waitSeconds: 0, queueCount: 0, matchedGameId: null, _idleRetries: 0 });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : 'Failed to leave queue' });
    }
  },

  pollStatus: async () => {
    if (get().status !== 'queued') return;
    try {
      const data = (await apiFetch('/api/v1/matchmaking/status')) as {
        status: MatchmakingStatus;
        waitSeconds?: number;
        queueCount?: number;
        gameId?: string;
      };

      if (data.status === 'matched' && data.gameId) {
        set({ status: 'matched', matchedGameId: data.gameId, _idleRetries: 0 });
      } else if (data.status === 'queued') {
        set({
          waitSeconds: data.waitSeconds || 0,
          queueCount: data.queueCount ?? get().queueCount,
          _idleRetries: 0,
        });
      } else {
        // Status returned 'idle' — might be a race condition between
        // queue removal and game creation. Don't immediately stop polling.
        const retries = get()._idleRetries + 1;
        if (retries >= 3) {
          // 3 consecutive idle responses (9s) — actually idle
          set({ status: 'idle', _idleRetries: 0 });
        } else {
          set({ _idleRetries: retries });
        }
      }
    } catch {
      // Silently fail individual polls
    }
  },

  reset: () => {
    set({ status: 'idle', waitSeconds: 0, queueCount: 0, matchedGameId: null, error: null, _idleRetries: 0 });
  },
}));

File: apps/Ios-app/Raichu/Raichu/Stores/MatchmakingStore.swift
// MatchmakingStore.swift
// Raichu
// ELO-based ranked matchmaking queue with 2-second polling.

import SwiftUI
import Combine

@MainActor
final class MatchmakingStore: ObservableObject {
    @Published var status: String = "idle"   // idle | queued | matched
    @Published var loading: Bool = false
    @Published var waitSeconds: Int = 0
    @Published var queueCount: Int = 0
    @Published var matchedGameId: String? = nil
    @Published var error: String? = nil

    private var idleRetries: Int = 0
    private var pollTask: Task<Void, Never>? = nil
    private var accessToken: String? = nil

    func joinQueue(accessToken: String) async {
        self.accessToken = accessToken
        loading = true
        error = nil
        defer { loading = false }

        do {
            _ = try await matchmakingAPI.joinQueue(accessToken: accessToken)
            status = "queued"
            idleRetries = 0
            startPolling()
        } catch {
            self.error = error.localizedDescription
        }
    }

    func leaveQueue() async {
        pollTask?.cancel()
        pollTask = nil
        guard let token = accessToken else { return }

        do {
            _ = try await matchmakingAPI.leaveQueue(accessToken: token)
        } catch { /* ignore leave errors */ }

        reset()
    }

    private func startPolling() {
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 2_000_000_000) // 2 seconds
                guard !Task.isCancelled else { break }
                await self?.pollStatus()
                if self?.status != "queued" { break }
            }
        }
    }

    func pollStatus() async {
        guard let token = accessToken else { return }

        do {
            let response = try await matchmakingAPI.status(accessToken: token)
            switch response.status {
            case "queued":
                status = "queued"
                waitSeconds = response.waitSeconds ?? 0
                queueCount = response.queueCount ?? 0
                idleRetries = 0
            case "matched":
                status = "matched"
                matchedGameId = response.gameId
                pollTask?.cancel()
            case "idle":
                idleRetries += 1
                // 3 consecutive idle responses → we're actually idle (race condition guard)
                if idleRetries >= 3 {
                    status = "idle"
                    pollTask?.cancel()
                }
            default:
                break
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func reset() {
        pollTask?.cancel()
        pollTask = nil
        status = "idle"
        loading = false
        waitSeconds = 0
        queueCount = 0
        matchedGameId = nil
        error = nil
        idleRetries = 0
        accessToken = nil
    }
}
